### PR TITLE
Updated sebastian/phpcpd, phpunit/phpunit and laminas/laminas-server dependency for php8 update

### PR DESCRIPTION
**Pull request by konarshankar07**
### Description (*)

This PR will update the packages related to php8 compatibility

### Fixed Issues (if relevant)

1. Fixes magento/magento2#31347
2. Fixes magento/magento2#31346
3. Fixes magento/magento2#31345
4. Fixes magento/magento2#31344
7. Fixes magento/magento2#31341
8. Fixes magento/magento2#31340
9. Fixes magento/magento2#31783

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sivaschenko** [comment]
@magento run all tests

File: composer.json
Comment (coderimus) [inline composer.json]: We must avoid using the dev as a `minimum-stability`. Please, change.

**Review by coderimus — CHANGES_REQUESTED** [review]
Hello @konarshankar07 
Thank you for the work you have done in the scope of the PHP8 compatibility. 👍

Please, take a look at my comment about the `minimum-stability`.

File: composer.json
Comment (sivaschenko) [inline composer.json]: I don't think the minimum stability should be dropped to dev. Is that required for some reason?

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the pull request @konarshankar07 ! Can you please take a look at my comment

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
Hello @sivaschenko  @coderimus 
Minimum stability is set to dev for the elasticsearch/elasticsearch update 
Thanks

**Comment by sivaschenko** [comment]
@konarshankar07 I see, elasticsearch 7.11 is not yet published. Let's update the elasticsearch version later, when 7.11 will be available on packagist https://packagist.org/packages/elasticsearch/elasticsearch to keep all the dependencies stable.

**Comment by konarshankar07** [comment]
@sivaschenko 
Should we downgrade the elasticsearch for now or we need to wait for the release?

**Comment by coderimus** [comment]
@konarshankar07 in our case we have to not to have any dev stability packages and taken into account the next 2 aspects we need to downgrade it to the stable one:
- 7.11 release will not be soon and, possibly, during this time we will have other packages updated and, maybe, new conflicts, so it is better to have PHP8 compatible system now :) to unblock Magento 3rd party vendors to upgrade their packages if needed.
- I think that the upgrade of the ES should be another complex task because ES is one of the critical packages which is responsible for data. It will need not only to upgrade version but also revied from the security perspective.
cc @sivaschenko

**Comment by konarshankar07** [comment]
@coderimus 
Thanks for the response. What do you suggest to move forward with this PR?

File: composer.json
Comment (sivaschenko) [inline composer.json]: Let's revert the elasticsearch version change

File: composer.json
Comment (sivaschenko) [inline composer.json]: this line should be removed

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @konarshankar07 , please see a summary of changes required for the PR to move forward

**Comment by konarshankar07** [comment]
@sivaschenko 
Thanks for the suggestion and I have reverted the elasticsearch update and removed the minimum stability.

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by coderimus** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Sample Data Tests CE , Static Tests, Unit Tests, WebAPI Tests

**Comment by coderimus** [comment]
@sivaschenko @konarshankar07 I have things to discuss:

1. I think we have to have PHP 8 instead of PHP 7.4 on our test builds. And we have to modify the composer `required section`: "php": "~7.3.0||~7.4.0|| ~8.0.0", 

image


2. I think we should raise a discussion about having the `laminas/laminas-code` 4.x.x in our project instead of the 3.x.x. From my perspective, having the 4.x.x is better because:
- 4.x.x will have support for all new PHP8 features
- It has PHP 7.4 as minimum supported version
- of course fixes and improvements according to the CHANGE.log
- having major version is good for future releases

This 4.0.0 version has some BC breaks and I think that they will not affect us or if so we will fix them :)

What do you think?

**Comment by konarshankar07** [comment]
Hello @coderimus 
yes, I agree with you but it should be covered in different PR and its part of https://github.com/magento/magento2/issues/31081 issue

**Comment by coderimus** [comment]
@konarshankar07 thank you for details and link PHP 8 upgrade issue 👍 Now I have a more clear vision of the whole process.

From my perspective, the PHP 8 version set to the required section of the composer.json should be done in the scope of this issue because without it we are not able to say that updated versions, Magento core code and 3rd party vendors included to the package are really compatible. Because here the PHP version is a conductor and dependencies are followers. Also, the `laminas/laminas-code` has a new version with fixes, updates for the PHP 8 which makes having 3.x.x version is not required, as for me.

I can suggest next:
- set PHP 8 requirement to the composer.json
- check result
- update `laminas/laminas-code` to the 4.x.x (fix BIC things)
- update all other libraries if not done in step 2

With this flow, I think, we will have PHP 8 compatibility upgrade consistently. Otherwise, we provide not the php8 dependency upgrade but the simple dependency upgrade :)

What do you think about this?

**Comment by sivaschenko** [comment]
Hi @coderimus @konarshankar07

We will not be able to run tests on PHP 8 until all dependencies are updated. Let's do this step by step.
I think it would be better to update laminas components to the minimal version compatible with PHP 8 in the scope of PHP 8 compatibility project, to avoid additional work and BIC.

**Comment by coderimus** [comment]
Hi @sivaschenko @konarshankar07 

@sivaschenko thank you for the update strategy explanation. Got your point and agreed with it. Let's proceed as you suggested.

**Comment by sivaschenko** [comment]
@konarshankar07 can you please check if the failed tests can be fixed.

I.e. it looks like lib/internal/Magento/Framework/composer.json should be updated as well to fix static tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by coderimus** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run Static Tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

**Pull request closed without merging**
